### DISCUSSION TITLE
Configuration deprecated packages dependency for NPM

### DISCUSSION THREAD

**Discussion by arushikesarwani94** [renovatebot/renovate#22578] [Suggest an Idea]
Trying to enable [Renovate Bot](https://github.com/renovatebot/renovate) in https://github.com/facebook/react-native to automatically create PRs to update package version when a NPM package dependency is deprecated simply by checking in a `renovate.json` configuration file in to [react-native](https://github.com/facebook/react-native)'s root directory.

Gone through the options mentioned in this guide : https://docs.renovatebot.com/configuration-options/ however couldn't find a suitable match for our use case. I have tried following but that isn't enough

```
    "enabledManagers": [
      "npm"
    ],
    "packageRules": [
      {
        "matchPackagePatterns": [
          "*@deprecated"
        ],
      }
    ],
```

It just matches patterns for the package name and not package info/metadata. Even with the `replacementName`, that's the same case.
> You can suggest a new community package rule by editing the [replacements.ts file on the Renovate repository](https://github.com/renovatebot/renovate/blob/main/lib/config/presets/internal/replacements.ts) and opening a pull request.

I would like to request a new feature that enables users to configure `renovate.json` for updating deprecated packages using RenovateBot when NPM is the package manager.

Thanks !

**Comment by HonkingGoose** [comment]
Heads up for the maintainers. This issue seems related to fetching metadata for a package in general:

- #11024

**Comment by rarkins** [comment]
`matchPackagePatterns` matches `packageName` (not metadata) and it uses regex. It looks like you're attempting to use a non-regex pattern and for metadata, so it's definitely not going to work if so.

Please take a step back and describe what you're trying to do? I'm not sure if you're updating dependencies, or replacing them. For example you mention "update package version" for deprecated but usually deprecated means there aren't anymore versions published.

**Reply by HonkingGoose** [reply]
### Update or replace dependencies?

> Please take a step back and describe what you're trying to do? I'm not sure if you're updating dependencies, or replacing them.

I think they want Renovate to _automatically_ update from deprecated package `deprecated-package` to replacement package `new-package`, based on npm's deprecation message for `deprecated-package`. Right now Renovate needs a Replacement Preset, or custom config from the user, to create a replacement PR.

#### npm does not have replacement metadata (that I can find)

According to the npm Docs, the package maintainer can enter whatever string they want for the deprecation message. "Whatever string they want" can't be parsed by Renovate automatically.

`npm-deprecate` can _not_ add metadata like: `deprecated-package` version `1.9.9` should be replaced with `new-package` version `2.0.0`. [^2] Relevant quote from the npm Docs: [^1]

> If you no longer wish to maintain a package, or if you would like to encourage users to update to a new or different version, you can [deprecate](https://docs.npmjs.com/cli/deprecate) it. Deprecating a package or version will print a message to the terminal when a user installs it.
>
> A deprecation warning or message can say anything. You may wish to include a message encouraging users to update to a specific version, or an alternate, supported package.

#### Raise warning in the Dependency Dashboard issue?

Looks like the best we could possibly do automatically, is to check each npm package that Renovate sees for a deprecation message. If Renovate sees a deprecation message, it's added to a list of deprecated packages which need user action.

> ### Deprecated packages (need user action)
>
> This package/these packages are deprecated, but Renovate can not create a replacement PR for them.
> You must replace these packages with a suitable replacement yourself:
>
> - `deprecated-package` version `1.9.9`

[^1]: [npm Docs, Deprecating and undeprecating packages or package versions](https://docs.npmjs.com/deprecating-and-undeprecating-packages-or-package-versions)
[^2]: [npm Docs, npm-deprecate](https://docs.npmjs.com/cli/v9/commands/npm-deprecate)

**Reply by arushikesarwani94** [reply]
Yes @HonkingGoose is correct. We want Renovate to automatically update from a deprecated package to suggested new replacement package whenever possible. Is there an existing configuration option in Renovate that can be used in order to check npm package for the deprecation message ? I can club that with `"dependencyDashboardApproval": true` in order to get it on [Dependency Dashboard](https://docs.renovatebot.com/key-concepts/dashboard/)

Even relying on community suggested replacement for deprecated packages would be meaningful.

> You can suggest a new community package rule by editing the [replacements.ts file on the Renovate repository](https://github.com/renovatebot/renovate/blob/main/lib/config/presets/internal/replacements.ts) and opening a pull request.

**Reply by rarkins** [reply]
A replacement feature exists already. Is there additional work required or is your need satisfied?

**Reply by arushikesarwani94** [reply]
@rarkins Replacement feature IIUC replaces a specified package name with community suggested ones. My question is let's say I set up the dependency dashboard, how do I configure it to use the npm package for the deprecation message just as in this one :
https://www.npmjs.com/package/uglify-es

**Reply by rarkins** [reply]
Sorry, I don't understand what you mean by "configure it to use the npm package for the deprecation message". Could you elaborate?

**Reply by HonkingGoose** [reply]
@rarkins they want Renovate to "read" the npm deprecation message and create the replacement PR automatically.

Right now you need to write code so Renovate knows how to replace a old package. I don't know if npm has bot-usable metadata that points to the replacement package.

The deprecation message from npm takes human input, so we can't reliably parse that.

**Reply by rarkins** [reply]
If that's the case then we can't do that - the npm data is not semantic/parseable enough for this to be done automatically

**Reply by arushikesarwani94** [reply]
@HonkingGoose Even if npm doesn't have a bot-usable metadata that points to the replacement package, it still has a metadata that points whether this is a deprecated package or not. My question is around how to use that information from npm through RenovateBot to surface the deprecated packages on the Renovate dependency dashboard ?
How do I make RenovateBot parse the npm metadata for deprecated package and surface it on [dependency dashboard](https://docs.renovatebot.com/key-concepts/dashboard/), suggesting replacement if found else not.

**Reply by rarkins** [reply]
- #5098
